File: apps/api/src/routes/sessions.ts
/** Session routes (SRS §6.5, §7.5): generate, list, update, roster. */
import { Hono } from "hono";
import { HTTPException } from "hono/http-exception";
import { generateSessionsSchema, updateSessionSchema, sessionListQuerySchema } from "@tuition/shared";
import type { AppContext, Db } from "../types";
import { parseBody, parseQuery } from "../lib/validate";
import { newId, nowIso } from "../lib/id";
import { writeAudit } from "../lib/db";
import { syncSessionToCalendar } from "../lib/google";
import { authenticate, requirePermission } from "../middleware/auth";

const SESSION_SELECT = `
  SELECT cs.id, cs.class_id, c.name AS class_name, c.code, c.band,
         cs.session_date, cs.start_time, cs.end_time, cs.status, cs.topic,
         cs.substitute_lecturer_id, cs.created_at,
         (SELECT COUNT(*) FROM enrollments e WHERE e.class_id = cs.class_id AND e.status = 'active') AS enrolled_count,
         (SELECT COUNT(*) FROM attendance a WHERE a.session_id = cs.id AND a.status IN ('present', 'late')) AS present_count
    FROM class_sessions cs
    JOIN classes c ON c.id = cs.class_id`;

const DAY_MS = 86_400_000;

async function getSession(db: Db, id: string) {
  return db.prepare(`${SESSION_SELECT} WHERE cs.id = ?`).bind(id).first();
}

export const sessionsRoutes = new Hono<AppContext>();
sessionsRoutes.use("*", authenticate);

/** POST /api/sessions/generate — materialize sessions from timetable slots. */
sessionsRoutes.post("/generate", requirePermission("session.manage"), async (c) => {
  const { class_id, from, to } = await parseBody(c, generateSessionsSchema);
  const db = c.get("db");

  const classRows = class_id
    ? await db.prepare(`SELECT id FROM classes WHERE id = ? AND deleted_at IS NULL AND status = 'active'`).bind(class_id).all<{ id: string }>()
    : await db.prepare(`SELECT id FROM classes WHERE deleted_at IS NULL AND status = 'active'`).all<{ id: string }>();
  const classes = classRows.results ?? [];
  if (classes.length === 0) return c.json({ created: 0, classes: 0 });

  const slotsByClass = new Map<string, { weekday: number; start_time: string; end_time: string }[]>();
  for (const cl of classes) {
    const slots = await db
      .prepare(`SELECT weekday, start_time, end_time FROM timetable_slots WHERE class_id = ? ORDER BY start_time`)
      .bind(cl.id)
      .all<{ weekday: number; start_time: string; end_time: string }>();
    slotsByClass.set(cl.id, slots.results ?? []);
  }

  const start = new Date(`${from}T00:00:00Z`).getTime();
  const end = new Date(`${to}T00:00:00Z`).getTime();
  if (Number.isNaN(start) || Number.isNaN(end)) throw new HTTPException(422, { message: "invalid_date" });
  if ((end - start) / DAY_MS > 366) throw new HTTPException(422, { message: "range_too_large" });

  const now = nowIso();
  const stmts: D1PreparedStatement[] = [];
  for (let t = start; t <= end; t += DAY_MS) {
    const d = new Date(t);
    const dateStr = d.toISOString().slice(0, 10);
    const wd = d.getUTCDay();
    for (const cl of classes) {
      const match = (slotsByClass.get(cl.id) ?? [])
        .filter((s) => s.weekday === wd)
        .sort((a, b) => a.start_time.localeCompare(b.start_time))[0];
      if (!match) continue;
      stmts.push(
        db
          .prepare(
            `INSERT OR IGNORE INTO class_sessions (id, class_id, session_date, start_time, end_time, status, created_at, updated_at)
             VALUES (?, ?, ?, ?, ?, 'scheduled', ?, ?)`,
          )
          .bind(newId("ses"), cl.id, dateStr, match.start_time, match.end_time, now, now),
      );
    }
  }

  let created = 0;
  if (stmts.length > 0) {
    const results = await db.batch(stmts);
    created = results.reduce((n, r) => n + (r.meta?.changes ?? 0), 0);
  }
  await writeAudit(db, {
    actorId: c.get("user").id,
    action: "session.generate",
    entityType: "class",
    entityId: class_id ?? null,
    after: { from, to, created },
  });
  return c.json({ created, classes: classes.length });
});

/** GET /api/sessions?from=&to=&class_id=&status= */
sessionsRoutes.get("/", requirePermission("session.read"), async (c) => {
  const { from, to, class_id, status } = parseQuery(c, sessionListQuerySchema);
  const db = c.get("db");
  const where: string[] = [];
  const binds: unknown[] = [];
  if (from) { where.push("cs.session_date >= ?"); binds.push(from); }
  if (to) { where.push("cs.session_date <= ?"); binds.push(to); }
  if (class_id) { where.push("cs.class_id = ?"); binds.push(class_id); }
  if (status) { where.push("cs.status = ?"); binds.push(status); }
  const whereSql = where.length > 0 ? ` WHERE ${where.join(" AND ")}` : "";
  const rows = await db
    .prepare(`${SESSION_SELECT}${whereSql} ORDER BY cs.session_date, cs.start_time`)
    .bind(...binds)
    .all();
  return c.json({ sessions: rows.results ?? [] });
});

/** GET /api/sessions/:id — single session. */
sessionsRoutes.get("/:id", requirePermission("session.read"), async (c) => {
  const session = await getSession(c.get("db"), c.req.param("id"));
  if (!session) throw new HTTPException(404, { message: "not_found" });
  return c.json(session);
});

/** GET /api/sessions/:id/roster — enrolled students + attendance state. */
sessionsRoutes.get("/:id/roster", requirePermission("session.read"), async (c) => {
  const db = c.get("db");
  const id = c.req.param("id");
  const session = await db.prepare(`SELECT id, class_id FROM class_sessions WHERE id = ?`).bind(id).first<{ id: string; class_id: string }>();
  if (!session) throw new HTTPException(404, { message: "not_found" });

  const rows = await db
    .prepare(
      `SELECT s.id, s.reg_no, s.full_name, s.phone, s.photo_url, s.status, s.card_status,
              a.status AS att_status, a.method, a.checked_in_at
         FROM enrollments e
         JOIN students s ON s.id = e.student_id AND s.deleted_at IS NULL
         LEFT JOIN attendance a ON a.session_id = ? AND a.student_id = s.id
        WHERE e.class_id = ? AND e.status = 'active'
        ORDER BY s.name_normalized`,
    )
    .bind(id, session.class_id)
    .all<Record<string, unknown>>();

  const roster = (rows.results ?? []).map((r) => ({
    student: {
      id: r.id,
      reg_no: r.reg_no,
      full_name: r.full_name,
      phone: r.phone,
      photo_url: r.photo_url,
      status: r.status,
      card_status: r.card_status,
    },
    status: r.att_status ?? null,
    method: r.method ?? null,
    checked_in_at: r.checked_in_at ?? null,
  }));
  return c.json({ session: await getSession(db, id), roster });
});

/** DELETE /api/sessions/:id/attendance/:studentId — clear a student's mark. */
sessionsRoutes.delete("/:id/attendance/:studentId", requirePermission("attendance.record"), async (c) => {
  const db = c.get("db");
  const id = c.req.param("id");
  const studentId = c.req.param("studentId");
  await db.prepare(`DELETE FROM attendance WHERE session_id = ? AND student_id = ?`).bind(id, studentId).run();
  await writeAudit(db, { actorId: c.get("user").id, action: "attendance.clear", entityType: "session", entityId: id, after: { student_id: studentId } });
  return c.json({ ok: true });
});

/** PATCH /api/sessions/:id — open/close/cancel, set topic / substitute. */
sessionsRoutes.patch("/:id", requirePermission("session.manage"), async (c) => {
  const body = await parseBody(c, updateSessionSchema);
  const db = c.get("db");
  const id = c.req.param("id");
  const existing = await db.prepare(`SELECT id FROM class_sessions WHERE id = ?`).bind(id).first<{ id: string }>();
  if (!existing) throw new HTTPException(404, { message: "not_found" });

  const sets: string[] = [];
  const binds: unknown[] = [];
  if (body.status !== undefined) { sets.push("status = ?"); binds.push(body.status); }
  if (body.topic !== undefined) { sets.push("topic = ?"); binds.push(body.topic ?? null); }
  if (body.substitute_lecturer_id !== undefined) { sets.push("substitute_lecturer_id = ?"); binds.push(body.substitute_lecturer_id ?? null); }
  sets.push("updated_at = ?"); binds.push(nowIso());
  await db.prepare(`UPDATE class_sessions SET ${sets.join(", ")} WHERE id = ?`).bind(...binds, id).run();
  await writeAudit(db, { actorId: c.get("user").id, action: "session.update", entityType: "session", entityId: id, after: body });
  // Best-effort Google Calendar sync (no-op unless connected); never blocks.
  c.executionCtx.waitUntil(syncSessionToCalendar(db, c.env, id));
  return c.json(await getSession(db, id));
});

File: apps/api/internal/sessions/sessions.go
// Package sessions serves class-session endpoints (session.* permissions):
// generate from the timetable, list/get, roster, attendance clear, and update.
package sessions

import (
	"context"
	"database/sql"
	"encoding/json"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"

	"attendly/api/internal/auth"
	"attendly/api/internal/cryptox"
	"attendly/api/internal/httpapi"
	"attendly/api/internal/store"
)

const sessionSelect = `
	SELECT cs.id, cs.class_id, c.name AS class_name, c.code, c.band,
	       cs.session_date, cs.start_time, cs.end_time, cs.status, cs.topic,
	       cs.substitute_lecturer_id, cs.created_at,
	       (SELECT COUNT(*) FROM enrollments e WHERE e.class_id = cs.class_id AND e.status = 'active') AS enrolled_count,
	       (SELECT COUNT(*) FROM attendance a WHERE a.session_id = cs.id AND a.status IN ('present', 'late')) AS present_count
	  FROM class_sessions cs JOIN classes c ON c.id = cs.class_id`

const dayMS = 24 * time.Hour

// SyncHook is an optional best-effort calendar sync invoked after a session
// changes; nil disables it.
type SyncHook func(sessionID string)

// Handlers serves the sessions domain.
type Handlers struct {
	db   *sql.DB
	sync SyncHook
}

// New constructs the sessions handlers. hook may be nil.
func New(db *sql.DB, hook SyncHook) *Handlers { return &Handlers{db: db, sync: hook} }

// Mount registers session routes (Authenticate applied upstream).
func (h *Handlers) Mount(r chi.Router) {
	read := auth.RequirePermission("session.read")
	manage := auth.RequirePermission("session.manage")
	r.Route("/api/sessions", func(r chi.Router) {
		r.With(manage).Method(http.MethodPost, "/generate", httpapi.Handler(h.generate))
		r.With(read).Method(http.MethodGet, "/", httpapi.Handler(h.list))
		r.With(read).Method(http.MethodGet, "/{id}", httpapi.Handler(h.get))
		r.With(read).Method(http.MethodGet, "/{id}/roster", httpapi.Handler(h.roster))
		r.With(auth.RequirePermission("attendance.record")).Method(http.MethodDelete, "/{id}/attendance/{studentId}", httpapi.Handler(h.clearAttendance))
		r.With(manage).Method(http.MethodPatch, "/{id}", httpapi.Handler(h.update))
	})
}

func (h *Handlers) actor(r *http.Request) string { return auth.MustUser(r.Context()).ID }

func (h *Handlers) getSession(ctx context.Context, id string) (store.Row, error) {
	return store.QueryFirstMap(ctx, h.db, sessionSelect+` WHERE cs.id = ?`, id)
}

type slot struct {
	weekday   int
	startTime string
	endTime   string
}

func (h *Handlers) generate(w http.ResponseWriter, r *http.Request) error {
	var in struct {
		ClassID string `json:"class_id"`
		From    string `json:"from"`
		To      string `json:"to"`
	}
	if err := httpapi.Decode(r, &in); err != nil {
		return err
	}
	start, err1 := time.Parse("2006-01-02", in.From)
	end, err2 := time.Parse("2006-01-02", in.To)
	if err1 != nil || err2 != nil {
		return httpapi.Err(http.StatusUnprocessableEntity, "invalid_date")
	}
	if end.Sub(start) > 366*dayMS {
		return httpapi.Err(http.StatusUnprocessableEntity, "range_too_large")
	}

	// Active classes in scope.
	var classQuery string
	args := []any{}
	if in.ClassID != "" {
		classQuery = `SELECT id FROM classes WHERE id = ? AND deleted_at IS NULL AND status = 'active'`
		args = append(args, in.ClassID)
	} else {
		classQuery = `SELECT id FROM classes WHERE deleted_at IS NULL AND status = 'active'`
	}
	classRows, err := h.db.QueryContext(r.Context(), classQuery, args...)
	if err != nil {
		return err
	}
	var classIDs []string
	for classRows.Next() {
		var id string
		if err := classRows.Scan(&id); err != nil {
			classRows.Close()
			return err
		}
		classIDs = append(classIDs, id)
	}
	classRows.Close()
	if err := classRows.Err(); err != nil {
		return err
	}
	if len(classIDs) == 0 {
		httpapi.JSON(w, http.StatusOK, map[string]any{"created": 0, "classes": 0})
		return nil
	}

	slotsByClass := map[string][]slot{}
	for _, cid := range classIDs {
		rows, err := h.db.QueryContext(r.Context(), `SELECT weekday, start_time, end_time FROM timetable_slots WHERE class_id = ? ORDER BY start_time`, cid)
		if err != nil {
			return err
		}
		for rows.Next() {
			var s slot
			if err := rows.Scan(&s.weekday, &s.startTime, &s.endTime); err != nil {
				rows.Close()
				return err
			}
			slotsByClass[cid] = append(slotsByClass[cid], s)
		}
		rows.Close()
	}

	now := cryptox.NowISO()
	created := 0
	tx, err := h.db.BeginTx(r.Context(), nil)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback() }()
	for t := start; !t.After(end); t = t.Add(dayMS) {
		dateStr := t.Format("2006-01-02")
		wd := int(t.Weekday())
		for _, cid := range classIDs {
			// earliest slot on this weekday
			var match *slot
			for i := range slotsByClass[cid] {
				s := slotsByClass[cid][i]
				if s.weekday == wd && (match == nil || s.startTime < match.startTime) {
					sc := s
					match = &sc
				}
			}
			if match == nil {
				continue
			}
			res, err := tx.ExecContext(r.Context(),
				`INSERT OR IGNORE INTO class_sessions (id, class_id, session_date, start_time, end_time, status, created_at, updated_at)
				 VALUES (?, ?, ?, ?, ?, 'scheduled', ?, ?)`,
				cryptox.NewID("ses"), cid, dateStr, match.startTime, match.endTime, now, now)
			if err != nil {
				return err
			}
			if n, _ := res.RowsAffected(); n > 0 {
				created++
			}
		}
	}
	if err := tx.Commit(); err != nil {
		return err
	}
	var classID *string
	if in.ClassID != "" {
		classID = &in.ClassID
	}
	actor := h.actor(r)
	_ = store.WriteAudit(r.Context(), h.db, store.AuditEntry{ActorID: &actor, Action: "session.generate", EntityType: sp("class"), EntityID: classID, After: map[string]any{"from": in.From, "to": in.To, "created": created}})
	httpapi.JSON(w, http.StatusOK, map[string]any{"created": created, "classes": len(classIDs)})
	return nil
}

func (h *Handlers) list(w http.ResponseWriter, r *http.Request) error {
	q := r.URL.Query()
	where, args := []string{}, []any{}
	if v := q.Get("from"); v != "" {
		where, args = append(where, "cs.session_date >= ?"), append(args, v)
	}
	if v := q.Get("to"); v != "" {
		where, args = append(where, "cs.session_date <= ?"), append(args, v)
	}
	if v := q.Get("class_id"); v != "" {
		where, args = append(where, "cs.class_id = ?"), append(args, v)
	}
	if v := q.Get("status"); v != "" {
		where, args = append(where, "cs.status = ?"), append(args, v)
	}
	whereSQL := ""
	if len(where) > 0 {
		whereSQL = " WHERE " + joinAnd(where)
	}
	var total int64
	if err := h.db.QueryRowContext(r.Context(), `SELECT COUNT(*) FROM class_sessions cs`+whereSQL, args...).Scan(&total); err != nil {
		return err
	}
	p := httpapi.ParsePage(r)
	rows, err := store.QueryMaps(r.Context(), h.db,
		sessionSelect+whereSQL+" ORDER BY cs.session_date, cs.start_time LIMIT ? OFFSET ?",
		append(args, p.Limit, p.Offset)...)
	if err != nil {
		return err
	}
	httpapi.JSON(w, http.StatusOK, map[string]any{"sessions": rows, "total": total, "page": p.Page, "page_size": p.PageSize})
	return nil
}

func (h *Handlers) get(w http.ResponseWriter, r *http.Request) error {
	row, err := h.getSession(r.Context(), chi.URLParam(r, "id"))
	if err != nil {
		return err
	}
	if row == nil {
		return httpapi.NotFound("not_found")
	}
	httpapi.JSON(w, http.StatusOK, row)
	return nil
}

func (h *Handlers) roster(w http.ResponseWriter, r *http.Request) error {
	id := chi.URLParam(r, "id")
	var classID string
	err := h.db.QueryRowContext(r.Context(), `SELECT class_id FROM class_sessions WHERE id = ?`, id).Scan(&classID)
	if err == sql.ErrNoRows {
		return httpapi.NotFound("not_found")
	}
	if err != nil {
		return err
	}
	rows, err := store.QueryMaps(r.Context(), h.db,
		`SELECT s.id, s.reg_no, s.full_name, s.phone, s.photo_url, s.status, s.card_status,
		        a.status AS att_status, a.method, a.checked_in_at
		   FROM enrollments e
		   JOIN students s ON s.id = e.student_id AND s.deleted_at IS NULL
		   LEFT JOIN attendance a ON a.session_id = ? AND a.student_id = s.id
		  WHERE e.class_id = ? AND e.status = 'active' ORDER BY s.name_normalized`, id, classID)
	if err != nil {
		return err
	}
	roster := make([]map[string]any, 0, len(rows))
	for _, r := range rows {
		roster = append(roster, map[string]any{
			"student": map[string]any{
				"id": r["id"], "reg_no": r["reg_no"], "full_name": r["full_name"],
				"phone": r["phone"], "photo_url": r["photo_url"], "status": r["status"], "card_status": r["card_status"],
			},
			"status": r["att_status"], "method": r["method"], "checked_in_at": r["checked_in_at"],
		})
	}
	sess, err := h.getSession(r.Context(), id)
	if err != nil {
		return err
	}
	httpapi.JSON(w, http.StatusOK, map[string]any{"session": sess, "roster": roster})
	return nil
}

func (h *Handlers) clearAttendance(w http.ResponseWriter, r *http.Request) error {
	id, studentID := chi.URLParam(r, "id"), chi.URLParam(r, "studentId")
	if _, err := h.db.ExecContext(r.Context(), `DELETE FROM attendance WHERE session_id = ? AND student_id = ?`, id, studentID); err != nil {
		return err
	}
	actor := h.actor(r)
	_ = store.WriteAudit(r.Context(), h.db, store.AuditEntry{ActorID: &actor, Action: "attendance.clear", EntityType: sp("session"), EntityID: &id, After: map[string]any{"student_id": studentID}})
	httpapi.JSON(w, http.StatusOK, map[string]any{"ok": true})
	return nil
}

func (h *Handlers) update(w http.ResponseWriter, r *http.Request) error {
	id := chi.URLParam(r, "id")
	var x string
	if err := h.db.QueryRowContext(r.Context(), `SELECT id FROM class_sessions WHERE id = ?`, id).Scan(&x); err == sql.ErrNoRows {
		return httpapi.NotFound("not_found")
	} else if err != nil {
		return err
	}
	var raw map[string]json.RawMessage
	if err := httpapi.Decode(r, &raw); err != nil {
		return err
	}
	sets, args := []string{}, []any{}
	for _, col := range []string{"status", "topic", "substitute_lecturer_id"} {
		if v, ok := raw[col]; ok {
			var s *string
			if err := json.Unmarshal(v, &s); err != nil {
				return httpapi.BadRequest("invalid_body")
			}
			sets, args = append(sets, col+" = ?"), append(args, s)
		}
	}
	sets = append(sets, "updated_at = ?")
	args = append(args, cryptox.NowISO())
	if _, err := h.db.ExecContext(r.Context(), `UPDATE class_sessions SET `+joinComma(sets)+` WHERE id = ?`, append(args, id)...); err != nil {
		return err
	}
	actor := h.actor(r)
	_ = store.WriteAudit(r.Context(), h.db, store.AuditEntry{ActorID: &actor, Action: "session.update", EntityType: sp("session"), EntityID: &id})
	if h.sync != nil {
		h.sync(id) // best-effort calendar sync (non-blocking)
	}
	row, err := h.getSession(r.Context(), id)
	if err != nil {
		return err
	}
	httpapi.JSON(w, http.StatusOK, row)
	return nil
}

func joinAnd(parts []string) string  { return joinWith(parts, " AND ") }
func joinComma(parts []string) string { return joinWith(parts, ", ") }
func joinWith(parts []string, sep string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += sep
		}
		out += p
	}
	return out
}

func sp(s string) *string { return &s }
